eks/cleanup.go

```diff
@@ -0,0 +1,415 @@
+
+			// Yay, we're detached, process the next network interface.
+			case isNIDetachedErr(niResult, err) || isNINotFoundErr(err):
+				logger.Infof("Network interface %s is detached.", aws.StringValue(ni.NetworkInterfaceId))
+				return nil // exit retry loop with success
+
+			default:
+				logger.Errorf("Error polling attachment for network interface %s", aws.StringValue(ni.NetworkInterfaceId))
+				return retry.FatalError{Underlying: err} // halt retries with error
+			}
+		})
+
+		// All the retries failed or we hit a fatal error.
+		if err != nil {
+			if isMaxRetriesExceededErr(err) {
+				return errors.WithStackTrace(NetworkInterfaceDetachedTimeoutError{aws.StringValue(ni.NetworkInterfaceId)})
+			}
+			return err
+		}
+	}
+
+	return nil
+}
+
+func waitForNetworkInterfacesToBeDeleted(
+	ec2Svc *ec2.EC2,
+	networkInterfaces []*ec2.NetworkInterface,
+	maxRetries int,
+	sleepBetweenRetries time.Duration,
+) error {
+	logger := logging.GetProjectLogger()
+
+	for _, ni := range networkInterfaces {
+		logger.Infof("Waiting for network interface %s to be deleted.", aws.StringValue(ni.NetworkInterfaceId))
+
+		// Poll for the new status
+		describeNetworkInterfacesInput := &ec2.DescribeNetworkInterfacesInput{
+			NetworkInterfaceIds: []*string{ni.NetworkInterfaceId},
+		}
+
+		err := retry.DoWithRetryE("Wait for Network Interface to be Deleted", maxRetries, sleepBetweenRetries, func() error {
+			_, err := ec2Svc.DescribeNetworkInterfaces(describeNetworkInterfacesInput)
+
+			switch {
+			// If there's no error, we have to keep trying.
+			case err == nil:
+				return errors.WithStackTrace(fmt.Errorf("Network Interface %s not deleted.", aws.StringValue(ni.NetworkInterfaceId))) // continue retrying
+
+			// Yay, it's deleted, process the next network interface.
+			case isNINotFoundErr(err):
+				logger.Infof("Network interface %s is deleted.", aws.StringValue(ni.NetworkInterfaceId))
+				return nil // exit retry loop with success
+
+			default:
+				return retry.FatalError{Underlying: err} // halt retries with error
+			}
+		})
+
+		// All the retries failed or we hit a fatal error.
+		if err != nil {
+			if isMaxRetriesExceededErr(err) {
+				return errors.WithStackTrace(NetworkInterfaceDeletedTimeoutError{aws.StringValue(ni.NetworkInterfaceId)})
+			}
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Used to look up the security group for the ALB ingress controller
+func lookupSecurityGroup(
+	ec2Svc *ec2.EC2,
+	vpcID string,
+	clusterID string,
+) (*ec2.DescribeSecurityGroupsOutput, error) {
+	logger := logging.GetProjectLogger()
+
+	logger.Infof("Looking up security group containing tag for EKS cluster %s", clusterID)
+	sgInput := &ec2.DescribeSecurityGroupsInput{
+		Filters: []*ec2.Filter{
+			{
+				Name:   aws.String("vpc-id"),
+				Values: []*string{aws.String(vpcID)},
+			},
+			{
+				// TODO: does this one still get created?
```

Comment:
Still TODO?

Reply:
@yorinasub17 I'm a little worried about this one. It must be that the tags changed on AWS's side of things. Users who have the older security group with that older tag might run into issues using this cleanup, because we currently aren't deleting that. This is also a departure from the original python script, which did look for the older tag, but not this newer one. What do you think?

Reply:
Unfortunately testing this case would not be possible anymore, since you'd need a cluster created before AWS changed tags on the SG. I wonder if we just list it as a known issue and move on.

Reply:
I wouldn't be too worried about this, as you need to update to the newer one for the loadbalancer controller to function when upgrading to https://github.com/gruntwork-io/terraform-aws-eks/releases/tag/v0.28.0

Reply:
Is it worth mentioning then that people should first upgrade to `0.29.1` before this release? That way they follow those manual steps to apply the new tags.

Reply:
I went ahead and mentioned this in the guide under known issues. https://github.com/gruntwork-io/terraform-aws-eks/blob/tf13/migration-guide.md#known-issues